File: python3/house_robber.py
"""
Space   : O(n)
Time    : O(n)
"""

class Solution:
    def rob(self, nums: List[int]) -> int:
        ans = 0
        leng = len(nums)
        dp = [0] * leng
        
        for idx in range(leng):
            if idx < 2:
                dp[idx] = nums[idx]
            elif idx == 2:
                dp[idx] = dp[idx-2] + nums[idx]
            else:
                dp[idx]  = max(dp[idx-2], dp[idx-3]) + nums[idx]
            ans = max(dp[idx], ans)
            
        return ans

File: golang/house_robber.go
package main

// Space   : O(n)
// Time    : O(n)

func rob1(nums []int) int {
	ans := 0
	leng := len(nums)
	dp := make([]int, leng)

	for i, v := range nums {
		if i == 0 || i == 1 {
			dp[i] = v
		} else if i == 2 {
			dp[i] = v + dp[i-2]
		} else {
			if dp[i-2] > dp[i-3] {
				dp[i] = v + dp[i-2]
			} else {
				dp[i] = v + dp[i-3]
			}
		}

		if ans < dp[i] {
			ans = dp[i]
		}

	}

	return ans
}
